homeassistant/components/media_player/_reproduce_state.py

```diff
@@ -0,0 +1,45 @@
+"""Module that groups code required to handle state restore for component."""
+from typing import Iterable
+
+from homeassistant.const import (
+    ATTR_STATE, SERVICE_MEDIA_PAUSE, SERVICE_MEDIA_PLAY, SERVICE_MEDIA_SEEK,
+    SERVICE_MEDIA_STOP, SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_VOLUME_MUTE,
+    SERVICE_VOLUME_SET, STATE_IDLE, STATE_OFF, STATE_ON, STATE_PAUSED,
+    STATE_PLAYING)
+from homeassistant.core import State
+from homeassistant.helpers.state import (
+    async_construct_service_schemas, async_reproduce_domain_state)
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.loader import bind_hass
+
+
+@bind_hass
+async def async_reproduce_states(hass: HomeAssistantType,
+                                 states: Iterable[State]) -> None:
+    """Reproduce component states."""
+    from . import (
+        SERVICE_PLAY_MEDIA,
+        SERVICE_SELECT_SOURCE,
+        SERVICE_SELECT_SOUND_MODE,
+        DOMAIN,
+    )
+
+    SERVICE_STATE_LIST = [
+        (SERVICE_TURN_ON, {ATTR_STATE: STATE_ON}),
+        (SERVICE_TURN_OFF, {ATTR_STATE: STATE_OFF}),
+        (SERVICE_MEDIA_PLAY, {ATTR_STATE: STATE_PLAYING}),
+        (SERVICE_MEDIA_STOP, {ATTR_STATE: STATE_IDLE}),
+        (SERVICE_MEDIA_PAUSE, {ATTR_STATE: STATE_PAUSED}),
+        (SERVICE_VOLUME_SET, {}),
+        (SERVICE_VOLUME_MUTE, {}),
+        (SERVICE_MEDIA_SEEK, {}),
+        (SERVICE_SELECT_SOURCE, {}),
+        (SERVICE_SELECT_SOUND_MODE, {}),
+        (SERVICE_PLAY_MEDIA, {}),
+    ]
+
+    service_schemas = async_construct_service_schemas(
```

Comment:
I am not a big fan of extracting it from the voluptuous schemas. Why can't we do something like:

```
if VOLUME in state.attributes:
    await hass.services.async_call(DOMAIN, SERVICE_VOLUME_SET, blocking=True, context=context)

# etc
```

Because I wouldn't want people to start limiting their architecture to have to support this abstraction.

Reply:
Works reasonable for services with single attributes. Gets messy for ones with multiple optional attributes like lights turn_on, set_temperature. You need extract the optional attributes somehow.

I can see the benefits of explicit calls, but it will grow codebase.

Reply:
```patch
diff --git a/homeassistant/components/media_player/_reproduce_state.py b/homeassistant/components/media_player/_reproduce_state.py
index 6f053c4..ab7acd4 100644
--- a/homeassistant/components/media_player/_reproduce_state.py
+++ b/homeassistant/components/media_player/_reproduce_state.py
@@ -1,45 +1,126 @@
 """Module that groups code required to handle state restore for component."""
+import asyncio
 from typing import Iterable
 
 from homeassistant.const import (
-    ATTR_STATE, SERVICE_MEDIA_PAUSE, SERVICE_MEDIA_PLAY, SERVICE_MEDIA_SEEK,
+    SERVICE_MEDIA_PAUSE, SERVICE_MEDIA_PLAY, SERVICE_MEDIA_SEEK,
     SERVICE_MEDIA_STOP, SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_VOLUME_MUTE,
     SERVICE_VOLUME_SET, STATE_IDLE, STATE_OFF, STATE_ON, STATE_PAUSED,
     STATE_PLAYING)
 from homeassistant.core import State
-from homeassistant.helpers.state import (
-    async_construct_service_schemas, async_reproduce_domain_state)
 from homeassistant.helpers.typing import HomeAssistantType
 from homeassistant.loader import bind_hass
 
 
-@bind_hass
-async def async_reproduce_states(hass: HomeAssistantType,
-                                 states: Iterable[State]) -> None:
+def generate_data_dict(entity_id, attributes = {}, keys = ()):
+    data = {}
+    data['entity_id'] = entity_id
+    for key in keys:
+        if key in attributes:
+            data[key] = attributes[key]
+    return data
+
+
+async def async_reproduce_state(hass: HomeAssistantType,
+                                state: State) -> None:
     """Reproduce component states."""
     from . import (
+        ATTR_MEDIA_VOLUME_LEVEL,
+        ATTR_MEDIA_VOLUME_MUTED,
+        ATTR_MEDIA_SEEK_POSITION,
+        ATTR_INPUT_SOURCE,
+        ATTR_SOUND_MODE,
+        ATTR_MEDIA_CONTENT_TYPE,
+        ATTR_MEDIA_CONTENT_ID,
+        ATTR_MEDIA_ENQUEUE,
         SERVICE_PLAY_MEDIA,
         SERVICE_SELECT_SOURCE,
         SERVICE_SELECT_SOUND_MODE,
         DOMAIN,
     )
 
-    SERVICE_STATE_LIST = [
-        (SERVICE_TURN_ON, {ATTR_STATE: STATE_ON}),
-        (SERVICE_TURN_OFF, {ATTR_STATE: STATE_OFF}),
-        (SERVICE_MEDIA_PLAY, {ATTR_STATE: STATE_PLAYING}),
-        (SERVICE_MEDIA_STOP, {ATTR_STATE: STATE_IDLE}),
-        (SERVICE_MEDIA_PAUSE, {ATTR_STATE: STATE_PAUSED}),
-        (SERVICE_VOLUME_SET, {}),
-        (SERVICE_VOLUME_MUTE, {}),
-        (SERVICE_MEDIA_SEEK, {}),
-        (SERVICE_SELECT_SOURCE, {}),
-        (SERVICE_SELECT_SOUND_MODE, {}),
-        (SERVICE_PLAY_MEDIA, {}),
-    ]
-
-    service_schemas = async_construct_service_schemas(
-        hass, DOMAIN, SERVICE_STATE_LIST)
-
-    await async_reproduce_domain_state(
-        hass, DOMAIN, states, service_schemas)
+    if state.state == STATE_ON:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_TURN_ON,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_OFF:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_TURN_OFF,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_PLAYING:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_PLAY,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_IDLE:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_STOP,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_PAUSED:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_PAUSE,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+
+    if ATTR_MEDIA_VOLUME_LEVEL in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_VOLUME_SET,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_VOLUME_LEVEL)),
+            blocking=True, context=state.context)
+
+    if ATTR_MEDIA_VOLUME_MUTED in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_VOLUME_MUTE,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_VOLUME_MUTED)),
+            blocking=True, context=state.context)
+
+    if ATTR_MEDIA_SEEK_POSITION in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_SEEK,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_SEEK_POSITION)),
+            blocking=True, context=state.context)
+
+    if ATTR_INPUT_SOURCE in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_SELECT_SOURCE,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_INPUT_SOURCE)),
+            blocking=True, context=state.context)
+
+    if ATTR_SOUND_MODE in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_SELECT_SOUND_MODE,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_SOUND_MODE)),
+            blocking=True, context=state.context)
+
+    if (ATTR_MEDIA_CONTENT_TYPE in state.attributes) and \
+       (ATTR_MEDIA_CONTENT_ID in state.attributes):
+        await hass.services.async_call(
+            DOMAIN, SERVICE_PLAY_MEDIA,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_CONTENT_TYPE,
+                                ATTR_MEDIA_CONTENT_ID,
+                                ATTR_MEDIA_ENQUEUE)),
+            blocking=True, context=state.context)
+
+
+@bind_hass
+async def async_reproduce_states(hass: HomeAssistantType,
+                                 states: Iterable[State]) -> None:
+
+    await asyncio.gather(*[
+        async_reproduce_state(hass, state)
+        for state in states])
```


This would be the equivalent explicit version. It's much easier for a typo to sneak in due to the amount of duplicated code. Easy to miss to extend should an attribute be added to a service.

Reply:
@balloob do you like this approach better? If so i'll update the pull request. We can tackle the cyclic imports once the architecture discussion has come to fruition. It's not too bad to import in the function for now imho.

Reply:
I like this approach better.